### PR TITLE
Sprint13 kl expansion

### LINKED ISSUE #135: Allow KL Expansion geometry to have custom number of expansion modes

### PR DESCRIPTION
**Pull request by amal-ghamdi**
closes #135

### CODE REVIEW

File: cuqi/geometry/_geometry.py [hunk 1]
```diff
@@ -650,17 +651,27 @@ class KLExpansion(Continuous1D):
...
+    num_modes : int, default None
+        Number of expansion modes to use in the KL expansion. If `num_modes` is 
+        None or larger than the number of grid points, all modes will be used.
+
     """
-    
+
     # init function defining parameters for the KL expansion
-    def __init__(self, grid,  decay_rate=2.5, normalizer=12.0, axis_labels=None, **kwargs):
+    def __init__(self, grid=None, decay_rate=2.5, normalizer=12.0, num_modes=None, axis_labels=None, **kwargs):
```
Comment (nabriis) [inline cuqi/geometry/_geometry.py:662]: Could you add doc to explain what happens when grid = None?

File: tests/test_geometry.py [hunk 2]
```diff
@@ -175,21 +175,119 @@ def test_stepExpansion_fun2par(projection, func):
...
 
-    assert np.allclose(f_geom, f_expected)
+    p[num_modes:] = 0
+    f_expected = _inverse_sin_discrete_transform_KL(
+        p, N, decay_rate, normalizer)
+
+    assert np.allclose(f_geom, f_expected) and\
+        len(geom.coefs) == geom.par_dim and\
+        geom.par_dim == geom.num_modes and\
+        len(geom.grid) == geom.fun_dim
```
Comment (nabriis) [inline tests/test_geometry.py:202]: I think seperate asserts will make the error easier to catch in future.

File: tests/test_geometry.py [hunk 3]
```diff
@@ -175,21 +175,119 @@ def test_stepExpansion_fun2par(projection, func):
...
+    # If num_modes > len(grid), num_modes is set to len(grid)
+    assert geom.num_modes == 100 and len(geom.coefs) == 100
+
+    # Update grid (num_modes > len(grid))
+    geom.grid = np.linspace(0, 1, 120)
+    assert geom.num_modes == 120 and len(geom.coefs) == 120
+
+    # Update grid (num_modes < len(grid))
+    geom.grid = np.linspace(0, 1, 300)
+    assert geom.num_modes == 200 and len(geom.coefs) == 200
```
Comment (nabriis) [inline tests/test_geometry.py:220]: Nice. Glad  you added tests. Now it makes sense :)

**Review by nabriis — APPROVED** [review]
Hi @amal-ghamdi, Nice job on this PR. I really like the tests and appreciate the effort put into the OO structure. Makes the KLExpansion quite powerful. I only a had a few minor comments, so I will just approve now.

File: cuqi/geometry/_geometry.py (hunk 1 above)
Comment (amal-ghamdi) [reply cuqi/geometry/_geometry.py:664]: I updated the documentation for grid.

File: tests/test_geometry.py (hunk 2 above)
Comment (amal-ghamdi) [reply tests/test_geometry.py]: 👍 done

**Comment by amal-ghamdi** [comment]
Thank you @nabriis for your helpful review!

File: cuqi/geometry/_geometry.py [hunk 4]
```diff
@@ -672,15 +685,49 @@ def normalizer(self):
 
     @property
     def coefs(self):
+        # Return None if self.num_modes is 0,
```
Comment (jakobsj) [inline cuqi/geometry/_geometry.py:688]: Before going to explain the special case of num_modes 0, can you please add a general description of what this does.

File: cuqi/geometry/_geometry.py [hunk 5]
```diff
@@ -672,15 +685,49 @@ def normalizer(self):
...
+        if self._coefs is None or len(self._coefs) != self.num_modes:
+            eigvals = np.array(range(1, self.par_dim+1))  # KL eigvals
+            self._coefs = 1/np.float_power(eigvals, self.decay_rate)
+
+        # Return the coefficients.
         return self._coefs
 
+    @property
+    def num_modes(self):
+        # If grid is not set, interpret grid dimension as 0.
```
Comment (jakobsj) [inline cuqi/geometry/_geometry.py:704]: Same for this one, please first explain what the method does in general before the special case.

**Review by jakobsj — APPROVED** [review]
Looks good - again just a couple of suggestions for improving documentation.

File: cuqi/geometry/_geometry.py (hunk 4 above)
Comment (amal-ghamdi) [reply cuqi/geometry/_geometry.py:690]: Good point, I added explanation.

File: cuqi/geometry/_geometry.py (hunk 5 above)
Comment (amal-ghamdi) [reply cuqi/geometry/_geometry.py:707]: here too

**Comment by amal-ghamdi** [comment]
Thank you @jakobsj for your helpful review and comments!

**Pull request merged**